File: controller/frames.py
"""Pure world-frame kinematics and target-frame conversion.

No function in this module reads MuJoCo state. Dynamic torso state and fixed
mount calibration are explicit inputs; arm FK/Jacobians come from Pinocchio.
"""

from pathlib import Path

import numpy as np
import pinocchio as pin

from controller.pin_fk import build_pin_model, pin_T_K_E
from controller.state import (
    ArmControllerState,
    DualArmControllerStates,
    DualArmFramedTargets,
    DualArmWorldTargets,
    FramedTarget,
    MountCalibration,
    PlantState,
    Pose,
    TargetFrame,
    Twist,
    WorldTarget,
)
from controller.transforms import pose_from_transform, transform_from_pose


_GEN3_PATH = (
    Path(__file__).resolve().parents[1]
    / "sim"
    / "assets"
    / "kinova_gen3"
    / "gen3.xml"
)
pin_model, pin_data, ee_frame_id = build_pin_model(
    _GEN3_PATH, "base_link", "pinch_site"
)


def compose_pose(parent_to_child, child_to_object):
    transform = (
        transform_from_pose(
            parent_to_child.position_m, parent_to_child.rotation
        )
        @ transform_from_pose(
            child_to_object.position_m, child_to_object.rotation
        )
    )
    return Pose(*pose_from_transform(transform))


def arm_controller_state(plant, side, calibration):
    """Derive one arm's world pose, twist, and Jacobian from explicit state."""
    if not isinstance(plant, PlantState):
        raise TypeError("plant must be a PlantState")
    if not isinstance(calibration, MountCalibration):
        raise TypeError("calibration must be a MountCalibration")

    joints = plant.arm(side)
    mount = calibration.for_arm(side)
    T_W_T = transform_from_pose(
        plant.torso_pose_world.position_m,
        plant.torso_pose_world.rotation,
    )
    T_T_B = transform_from_pose(mount.position_m, mount.rotation)
    T_B_E = pin_T_K_E(
        pin_model, pin_data, ee_frame_id, joints.position_rad
    )
    ee_pose_world = Pose(*pose_from_transform(T_W_T @ T_T_B @ T_B_E))

    J_B = pin.computeFrameJacobian(
        pin_model,
        pin_data,
        joints.position_rad,
        ee_frame_id,
        pin.LOCAL_WORLD_ALIGNED,
    )
    R_W_B = plant.torso_pose_world.rotation @ mount.rotation
    jacobian_world = np.empty_like(J_B)
    jacobian_world[:3] = R_W_B @ J_B[:3]
    jacobian_world[3:] = R_W_B @ J_B[3:]

    arm_twist = jacobian_world @ joints.velocity_rad_s
    torso_twist = plant.torso_twist_world
    offset_world = (
        ee_pose_world.position_m - plant.torso_pose_world.position_m
    )
    ee_twist_world = Twist(
        torso_twist.linear_m_s
        + np.cross(torso_twist.angular_rad_s, offset_world)
        + arm_twist[:3],
        torso_twist.angular_rad_s + arm_twist[3:],
    )
    return ArmControllerState(
        joints=joints,
        ee_pose_world=ee_pose_world,
        ee_twist_world=ee_twist_world,
        jacobian_world=jacobian_world,
    )


def controller_states(plant, calibration):
    return DualArmControllerStates(
        right=arm_controller_state(plant, "right", calibration),
        left=arm_controller_state(plant, "left", calibration),
    )


def _world_reference_frame(plant, side, calibration, frame):
    """Return world pose/twist of the selected reference frame."""
    if frame == TargetFrame.WORLD:
        return Pose(np.zeros(3), np.eye(3)), Twist.zero()
    if frame == TargetFrame.TORSO:
        return plant.torso_pose_world, plant.torso_twist_world
    if frame == TargetFrame.BASE:
        mount = calibration.for_arm(side)
        pose_world = compose_pose(plant.torso_pose_world, mount)
        offset_world = (
            pose_world.position_m - plant.torso_pose_world.position_m
        )
        torso_twist = plant.torso_twist_world
        twist_world = Twist(
            torso_twist.linear_m_s
            + np.cross(torso_twist.angular_rad_s, offset_world),
            torso_twist.angular_rad_s,
        )
        return pose_world, twist_world
    raise ValueError(f"unsupported target frame: {frame!r}")


def resolve_target_world(plant, side, calibration, target):
    """Convert a world/base/torso target once at the frame boundary.

    Controllers receive only the returned ``WorldTarget`` and never see the
    frame selector.
    """
    if not isinstance(plant, PlantState):
        raise TypeError("plant must be a PlantState")
    if not isinstance(calibration, MountCalibration):
        raise TypeError("calibration must be a MountCalibration")
    if not isinstance(target, FramedTarget):
        raise TypeError("target must be a FramedTarget")

    frame_pose, frame_twist = _world_reference_frame(
        plant, side, calibration, target.reference_frame
    )
    pose_world = compose_pose(frame_pose, target.pose)
    offset_world = frame_pose.rotation @ target.pose.position_m
    twist_world = Twist(
        frame_twist.linear_m_s
        + np.cross(frame_twist.angular_rad_s, offset_world)
        + frame_pose.rotation @ target.twist.linear_m_s,
        frame_twist.angular_rad_s
        + frame_pose.rotation @ target.twist.angular_rad_s,
    )
    return WorldTarget(pose_world, twist_world)


def resolve_targets_world(plant, calibration, targets):
    if not isinstance(targets, DualArmFramedTargets):
        raise TypeError("targets must be DualArmFramedTargets")
    return DualArmWorldTargets(
        right=resolve_target_world(
            plant, "right", calibration, targets.right),
        left=resolve_target_world(
            plant, "left", calibration, targets.left),
    )

File: cpp/src/kinematics/Frames.cpp
#include "kinematics/Frames.h"

#include "math/Transforms.h"

namespace srl::kinematics {
namespace {

// World pose/twist of the selected reference frame (frames._world_reference_frame).
std::pair<Pose, Twist> WorldReferenceFrame(const PlantState& plant, Side side,
                                           const MountCalibration& calibration,
                                           TargetFrame frame) {
  switch (frame) {
    case TargetFrame::World:
      return {Pose{}, Twist::Zero()};
    case TargetFrame::Torso:
      return {plant.torso_pose_world, plant.torso_twist_world};
    case TargetFrame::Base: {
      const Pose& mount = calibration.for_arm(side);
      const Pose pose_world = ComposePose(plant.torso_pose_world, mount);
      const Eigen::Vector3d offset_world =
          pose_world.position_m - plant.torso_pose_world.position_m;
      const Twist& torso = plant.torso_twist_world;
      Twist twist_world;
      twist_world.linear_m_s =
          torso.linear_m_s + torso.angular_rad_s.cross(offset_world);
      twist_world.angular_rad_s = torso.angular_rad_s;
      return {pose_world, twist_world};
    }
  }
  throw std::invalid_argument("unsupported target frame");
}

}  // namespace

Pose ComposePose(const Pose& parent_to_child, const Pose& child_to_object) {
  const Eigen::Matrix4d transform =
      transforms::TransformFromPose(parent_to_child.position_m,
                                    parent_to_child.rotation) *
      transforms::TransformFromPose(child_to_object.position_m,
                                    child_to_object.rotation);
  Pose pose;
  transforms::PoseFromTransform(transform, pose.position_m, pose.rotation);
  return pose;
}

ArmControllerState ArmControllerStateOf(PinModel& pin, const PlantState& plant,
                                        Side side,
                                        const MountCalibration& calibration) {
  const ArmJointState& joints = plant.arm(side);
  const Pose& mount = calibration.for_arm(side);

  const Eigen::Matrix4d transform_world_torso = transforms::TransformFromPose(
      plant.torso_pose_world.position_m, plant.torso_pose_world.rotation);
  const Eigen::Matrix4d transform_torso_base =
      transforms::TransformFromPose(mount.position_m, mount.rotation);
  const Eigen::Matrix4d transform_world_base =
      transform_world_torso * transform_torso_base;

  pin.Update(joints.position_rad);
  const Eigen::Matrix4d transform_base_ee = pin.EeTransform();

  ArmControllerState state;
  state.joints = joints;
  transforms::PoseFromTransform(transform_world_base * transform_base_ee,
                                state.ee_pose_world.position_m,
                                state.ee_pose_world.rotation);

  const Matrix6x7 jacobian_base = pin.EeJacobian();
  // Computed as R_W_T * R_T_B rather than reused from the 4x4 product, exactly
  // as the Python does -- the two agree mathematically but not bit-for-bit.
  const Eigen::Matrix3d rotation_world_base =
      plant.torso_pose_world.rotation * mount.rotation;
  state.jacobian_world.topRows<3>() =
      rotation_world_base * jacobian_base.topRows<3>();
  state.jacobian_world.bottomRows<3>() =
      rotation_world_base * jacobian_base.bottomRows<3>();

  const Vector6 arm_twist = state.jacobian_world * joints.velocity_rad_s;
  const Twist& torso_twist = plant.torso_twist_world;
  const Eigen::Vector3d offset_world =
      state.ee_pose_world.position_m - plant.torso_pose_world.position_m;
  state.ee_twist_world.linear_m_s = torso_twist.linear_m_s +
                                    torso_twist.angular_rad_s.cross(offset_world) +
                                    arm_twist.head<3>();
  state.ee_twist_world.angular_rad_s =
      torso_twist.angular_rad_s + arm_twist.tail<3>();

  pin.ResolveLinkSafetyPoints(transform_world_base.topRightCorner<3, 1>(),
                              rotation_world_base, state.link_safety_points);
  return state;
}

DualArmControllerStates ControllerStates(PinModel& pin, const PlantState& plant,
                                         const MountCalibration& calibration) {
  DualArmControllerStates states;
  states.right = ArmControllerStateOf(pin, plant, Side::Right, calibration);
  states.left = ArmControllerStateOf(pin, plant, Side::Left, calibration);
  return states;
}

Pose ExpressPoseInTargetFrame(const PlantState& plant, Side side,
                              const MountCalibration& calibration,
                              TargetFrame frame, const Pose& pose_world) {
  const auto [frame_pose_world, unused_twist] =
      WorldReferenceFrame(plant, side, calibration, frame);
  (void)unused_twist;
  const Eigen::Matrix3d rotation_frame_world = frame_pose_world.rotation.transpose();
  return Pose(
      rotation_frame_world *
          (pose_world.position_m - frame_pose_world.position_m),
      rotation_frame_world * pose_world.rotation);
}

WorldTarget ResolveTargetWorld(const PlantState& plant, Side side,
                               const MountCalibration& calibration,
                               const FramedTarget& target) {
  const auto [frame_pose, frame_twist] =
      WorldReferenceFrame(plant, side, calibration, target.reference_frame);

  WorldTarget resolved;
  resolved.pose_world = ComposePose(frame_pose, target.pose);
  const Eigen::Vector3d offset_world =
      frame_pose.rotation * target.pose.position_m;
  resolved.twist_world.linear_m_s =
      frame_twist.linear_m_s + frame_twist.angular_rad_s.cross(offset_world) +
      frame_pose.rotation * target.twist.linear_m_s;
  resolved.twist_world.angular_rad_s =
      frame_twist.angular_rad_s + frame_pose.rotation * target.twist.angular_rad_s;
  return resolved;
}

DualArmWorldTargets ResolveTargetsWorld(const PlantState& plant,
                                        const MountCalibration& calibration,
                                        const DualArmFramedTargets& targets) {
  DualArmWorldTargets resolved;
  resolved.right =
      ResolveTargetWorld(plant, Side::Right, calibration, targets.right);
  resolved.left =
      ResolveTargetWorld(plant, Side::Left, calibration, targets.left);
  return resolved;
}

}  // namespace srl::kinematics
